mlkem/poly.c

```diff
@@ -274,28 +274,38 @@ void poly_frombytes(poly *r, const uint8_t a[MLKEM_POLYBYTES]) {
 }
 #endif /* MLKEM_USE_NATIVE_POLY_FROMBYTES */
 
-/*************************************************
- * Name:        poly_frommsg
- *
- * Description: Convert 32-byte message to polynomial
- *
- * Arguments:   - poly *r: pointer to output polynomial
- *              - const uint8_t *msg: pointer to input message
- **************************************************/
 void poly_frommsg(poly *r, const uint8_t msg[MLKEM_INDCPA_MSGBYTES]) {
-  unsigned int i, j;
-
 #if (MLKEM_INDCPA_MSGBYTES != MLKEM_N / 8)
 #error "MLKEM_INDCPA_MSGBYTES must be equal to MLKEM_N/8 bytes!"
 #endif
 
-  for (i = 0; i < MLKEM_N / 8; i++) {
-    for (j = 0; j < 8; j++) {
-      r->coeffs[8 * i + j] = 0;
-      cmov_int16(r->coeffs + 8 * i + j, ((MLKEM_Q + 1) / 2), (msg[i] >> j) & 1);
-    }
-  }
+  const int num_blocks = MLKEM_N / 8;
+
+  for (int i = 0; i < num_blocks; i++)
+      // clang-format off
+    ASSIGNS(i, OBJECT_WHOLE(r))
+    INVARIANT(i >= 0)
+    INVARIANT(i <= num_blocks)
+    INVARIANT(num_blocks == 32)
```

Comment:
This should not be needed?

Reply:
Removing that num_blocks == 32 invariants changes the time to prove this function (1 CPU on my laptop) from 3 min 47 seconds to 5 min 20 seconds. Worth removing or not?

Reply:
Can we not use write `MLKEM_N / 8` everywhere? That's how it's done in the other loops.

Reply:
I took the liberty to change that.

Reply:
OK. I personally find the repetition dubious, but happy to oblige. I guess the repetition is only risky if there's a chance the value might change, which is in this case is zero...